data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/validator/PostContributionAndProofs.java

```diff
@@ -74,31 +93,39 @@ public PostContributionAndProofs(
       })
   @Override
   public void handle(@NotNull final Context ctx) throws Exception {
-    final SignedContributionAndProof[] signedContributionAndProofs;
-    try {
-      signedContributionAndProofs =
-          parseRequestBody(ctx.body(), SignedContributionAndProof[].class);
-    } catch (IllegalArgumentException e) {
-      ctx.json(BadRequest.badRequest(jsonProvider, e.getMessage()));
-      ctx.status(HttpStatusCodes.SC_BAD_REQUEST);
-      return;
-    }
-    final SafeFuture<Void> future =
-        provider.sendContributionAndProofs(asList(signedContributionAndProofs));
+    adapt(ctx);
+  }
 
-    ctx.future(
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final List<SignedContributionAndProof> signedContributionAndProofs = request.getRequestBody();
+    final SafeFuture<Void> future = provider.sendContributionAndProofs(signedContributionAndProofs);
+
+    request.respondAsync(
         future
-            .thenApplyChecked(result -> "")
-            .exceptionallyCompose(error -> handleError(ctx, error)));
+            .thenApply(v -> AsyncApiResponse.respondWithCode(SC_OK))
+            .exceptionallyCompose(
+                error -> {
+                  final Throwable rootCause = Throwables.getRootCause(error);
+                  if (rootCause instanceof IllegalArgumentException) {
+                    return SafeFuture.of(
+                        () ->
+                            AsyncApiResponse.respondWithError(SC_BAD_REQUEST, error.getMessage()));
+                  }
+                  return SafeFuture.of(
+                      () ->
+                          AsyncApiResponse.respondWithError(
+                              SC_BAD_REQUEST, "Failed to submit contribution and proofs"));
+                }));
```

Comment:
the illegal arg should be automatically handled, not sure about the second bit... would be interesting to see if its still the same result without the exceptionally...

Reply:
I'm thinking the whole `exceptionallyCompose` is probably not needed

Reply:
I'll have to have a play, didnt get to it today but I'll look tomorrow morning :)

Reply:
- validatorDataProvider will call either remoteValidatorApiHandler or ValidatorApiHandler
 - validatorApiHandler will throw IllegalArg, anything else is likely a bug
 - remoteValidatorApiHandler will decode then put onto validatorApiHandler, so will be JsonProcessingException or IllegalArg...
based on this, we're only really expecting an IllegalArgumentException or JsonProcessingException, and these are both handled by the generic exception handler.
If the old behaviour of an unexpected exception coming out were to actually happen, we'd potentially be missing useful information by swallowing it in the old way, so i think yes, remove the ExceptionallyCompose.
If we do see any 500 errors, we'll need to accept that tech debt, but whether 400 or 500, the answer is still that the POST failed.